docs/adrs/0003-profile-layering.md

```diff
@@ -0,0 +1,119 @@
+# 3. Profile Layering
+
+Date: 2021-11-16
+
+## Status
+
+Accepted
+
+## Context
+
+There are cases where we require to install profiles in a specific order,
+perhaps the components being installed are required to be functional before
+further workloads are installed into the cluster.
+
+Flux HelmReleases support this through the [`dependsOn` mechanism](https://fluxcd.io/docs/components/helm/helmreleases/#helmrelease-dependencies).
+
+Because the `HelmReleases` are generated inside the CAPI service, we want to
+provide a way to configure this dependency without having to manually edit the
+files.
+
+We may not always know the specific dependency to be able to set an explicit
+dependency, for example, "certificate management" should be in place before
+"service mesh".
+
+When installing a set of profiles, it should be possible to declaratively say
+that some profiles should be installed before others if this is required.
+
+## Decision
+
+Support Helm Chart [annotation](https://helm.sh/docs/topics/charts/#the-chartyaml-file) that indicates a layer-like ordering for unrelated profiles.
+
+The `weave.works/layer` indicates that a Profile chart should be applied in a
+specific layer.
+
+For example, with the following two Profile charts.
```

Comment:
Please provide a more elaborate configuration. Like 10 profiles at least so we can see how complex and convoluted this gets and how fast...

Reply:
The nice thing is that it doesn't get much more complicated...certainly simpler than explicit dependencies.

Reply:
For sure, but can we nest layers, and how would that look like in a bigger configuration?

Reply:
I mean I guess nested layering is a noop since a has layer a and b has layer b and c can have layer b then since tha already has layer a.

Reply:
But what about multiple layers which aren't related? So C has layer a and layer b which do not co-depend?

Reply:
I think you're misunderstanding the layering, which means I need to do some work to clarify in the document.

The layers are for ordering of dependencies, and not for explicitly defining layers.

Reply:
Oh. :)

Reply:
Ah, okay, it's clear now. The ordering of the dependencies is happening using the lexically sorted layer names. The feature is the name of the layer. :) Gotcha.